### ISSUE TITLE
Emit diagnostic on abstract classes that don't directly reference ABC/ABCMeta

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ty#2686]
Basically just a ty version of this basedpyright rule:
https://docs.basedpyright.com/v1.37.2/benefits-over-pyright/new-diagnostic-rules/#reportimplicitabstractclass

In addition to all the benefits listed, my favorite part even on non-library code is how it keeps the diagnostic local to the class you are writing - especially when I'm sketching out the class structure for a project, it can be a while before I write any code that actually constructs instances of an ABC, which by then the finer ideas about implementing a class can be forgotten. Having the error for a maybe-not-supposed-to-be-abstract class be up-front on the class itself allows me to fix it on the spot with the details fresh in my mind.

Example from the basedpyright docs:
https://play.ty.dev/6036ee7f-e115-41f0-a117-2aad30be7e04
```py
from abc import ABC, abstractmethod

class AbstractFoo(ABC):
    @abstractmethod
    def foo(self):
        ...

class FooImpl(AbstractFoo):  # Add new diagnostic here for implicitly abstract class
    def bar(self):
        print("hi")

foo = FooImpl()
```

Depending on the ecosystem impact it may need to be ignored by default

**Comment by carljm** [comment]
Hmm. Currently ty respects `@abstractmethod` as a type-level indication that a method should be considered abstract, and errors on a final subclass that doesn't override the method, regardless of whether even the base class inherits `ABC`. That seems somewhat incompatible with the proposed check.

**Comment by MeGaGiGaGon** [comment]
Interesting, I wasn't aware of that. On thinking more, maybe then this would be better suited for the hypothetical ruff + ty fully type aware linter? Since it could be considered more of a "you might be doing this wrong" issue, instead of a "this is a type error" issue.
Edit: Or maybe it could even be a ruff lint right now? I recall that the type knowledge of ruff is currently limited to a single file, but at least for me that would already catch all of my use case, and I think you could get up to a solid "most" if you also included a default list of standard library abstract classes.

**Comment by AlexWaygood** [comment]
See https://github.com/astral-sh/ruff/issues/12861 for previous discussion on the ruff side. I wouldn't be opposed to adding this as a disabled-by-default lint to ty now, since I agree it would help some people and it requires multifile analysis to do it effectively. But as Carl says, we don't require the metaclass in ty in order for us to consider a method abstract at type-checking time, and that's very intentional. So I don't think it should be enabled by default.

**Comment by MeGaGiGaGon** [comment]
https://github.com/astral-sh/ruff/issues/12861 is similar, but not exactly the same - it's talking about when a class has methods with explicit abstract decorators on it, whereas this would be for when you end up with implicitly abstract methods on a seemingly normal class.
Edit: On reading the source code it looks like I am actually wrong, it does call out this in the docs on the rule in the PR
```rust
/// For a mixin class, it is not enough that `abc.ABC` is included in the eventual MRO.
/// The mixin class must also inherit directly from `ABC` for the decorators to take effect.
```

**Comment by AlexWaygood** [comment]
> [astral-sh/ruff#12861](https://github.com/astral-sh/ruff/issues/12861) is similar, but not exactly the same - it's talking about when a class has methods with explicit abstract decorators on it, whereas this would be for when you end up with implicitly abstract methods on a seemingly normal class. Edit: On reading the source code it looks like I am actually wrong, it does call out this in the docs on the rule in the PR

Oh I see. That would also be very strict, but I can see the use case. Mypy actually has a version of that check that it has enabled by default, but only for stub files, because it's much easier to make a class accidentally implicitly abstract in a stub file if you forget to override an abstract method in a subclass. It's caught some bugs over at typeshed in the past, but it's also often been annoying and had quite a few false positives.
